File: articles/mathematical-modeling-in-artificial-intelligence-and-data-systems/fortran/ai_candidate_governance_summary.f90
program ai_candidate_governance_summary
  implicit none

  integer, parameter :: n = 4
  character(len=32) :: keys(n)
  real(8) :: validation(n), calibration(n), subgroup_gap(n), drift(n), interpretability(n), privacy(n), criticality(n)
  real(8) :: score(n)
  integer :: i
  logical :: review

  keys = (/ 'baseline_logistic              ', 'tree_ensemble                  ', &
            'neural_model                   ', 'constrained_model              ' /)
  validation = (/ 0.76d0, 0.83d0, 0.86d0, 0.81d0 /)
  calibration = (/ 0.050d0, 0.070d0, 0.095d0, 0.035d0 /)
  subgroup_gap = (/ 0.080d0, 0.140d0, 0.190d0, 0.060d0 /)
  drift = (/ 0.120d0, 0.180d0, 0.240d0, 0.100d0 /)
  interpretability = (/ 0.920d0, 0.620d0, 0.380d0, 0.780d0 /)
  privacy = (/ 0.080d0, 0.130d0, 0.180d0, 0.090d0 /)
  criticality = (/ 0.620d0, 0.700d0, 0.820d0, 0.660d0 /)

  print '(A)', 'key validation_score calibration_error subgroup_error_gap drift_score privacy_risk governance_score requires_review'

  do i = 1, n
    score(i) = governance_score(validation(i), calibration(i), subgroup_gap(i), drift(i), interpretability(i), privacy(i), criticality(i))
    review = calibration(i) > 0.08d0 .or. subgroup_gap(i) > 0.12d0 .or. drift(i) > 0.20d0 .or. privacy(i) > 0.15d0 .or. interpretability(i) < 0.50d0
    print '(A,1X,F8.4,1X,F8.4,1X,F8.4,1X,F8.4,1X,F8.4,1X,F10.4,1X,L1)', &
      trim(keys(i)), validation(i), calibration(i), subgroup_gap(i), drift(i), privacy(i), score(i), review
  end do

contains

  real(8) function governance_score(validation, calibration, subgroup_gap, drift, interpretability, privacy, criticality)
    implicit none
    real(8), intent(in) :: validation, calibration, subgroup_gap, drift, interpretability, privacy, criticality
    governance_score = validation - (1.8d0 * calibration + 1.5d0 * subgroup_gap + 1.2d0 * drift + 1.4d0 * privacy + 0.7d0 * criticality - 0.5d0 * interpretability)
  end function governance_score

end program ai_candidate_governance_summary

File: articles/mathematical-modeling-in-artificial-intelligence-and-data-systems/c/ai_candidate_governance_summary.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    const char *key;
    const char *model_name;
    double validation_score;
    double calibration_error;
    double subgroup_error_gap;
    double drift_score;
    double interpretability_score;
    double privacy_risk;
    double deployment_criticality;
} ModelCandidate;

double governance_score(ModelCandidate candidate) {
    double penalty =
        1.8 * candidate.calibration_error +
        1.5 * candidate.subgroup_error_gap +
        1.2 * candidate.drift_score +
        1.4 * candidate.privacy_risk +
        0.7 * candidate.deployment_criticality -
        0.5 * candidate.interpretability_score;
    return candidate.validation_score - penalty;
}

int requires_review(ModelCandidate candidate) {
    return candidate.calibration_error > 0.08 ||
           candidate.subgroup_error_gap > 0.12 ||
           candidate.drift_score > 0.20 ||
           candidate.privacy_risk > 0.15 ||
           candidate.interpretability_score < 0.50;
}

int main(void) {
    ModelCandidate candidates[] = {
        {"baseline_logistic", "Baseline logistic model", 0.76, 0.050, 0.080, 0.120, 0.920, 0.080, 0.62},
        {"tree_ensemble", "Tree ensemble", 0.83, 0.070, 0.140, 0.180, 0.620, 0.130, 0.70},
        {"neural_model", "Neural model", 0.86, 0.095, 0.190, 0.240, 0.380, 0.180, 0.82},
        {"constrained_model", "Constrained calibrated model", 0.81, 0.035, 0.060, 0.100, 0.780, 0.090, 0.66}
    };

    printf("key,validation_score,calibration_error,subgroup_error_gap,drift_score,privacy_risk,governance_score,requires_review\n");

    for (int i = 0; i < 4; ++i) {
        printf("%s,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%s\n",
               candidates[i].key,
               candidates[i].validation_score,
               candidates[i].calibration_error,
               candidates[i].subgroup_error_gap,
               candidates[i].drift_score,
               candidates[i].privacy_risk,
               governance_score(candidates[i]),
               requires_review(candidates[i]) ? "true" : "false");
    }

    fprintf(stderr, "ai_candidate_governance_summary complete\n");
    return EXIT_SUCCESS;
}
